sys/kern/vfs_aio.c

```diff
@@ -2991,4 +3056,305 @@ freebsd32_lio_listio(struct thread *td, struct freebsd32_lio_listio_args *uap)
 	return (error);
 }
 
-#endif
+#endif /* COMPAT_FREEBSD32 */
+
+#ifdef COMPAT_CHERIABI
+#include <sys/user.h>
+
+#include <machine/cherireg.h>
+
+#include <cheri/cheric.h>
+
+#include <compat/cheriabi/cheriabi.h>
+#include <compat/cheriabi/cheriabi_proto.h>
+#include <compat/cheriabi/cheriabi_signal.h>
+#include <compat/cheriabi/cheriabi_syscall.h>
+#include <compat/cheriabi/cheriabi_util.h>
+
+struct __aiocb_private_c {
+	uint64_t	status;
+	uint64_t	error;
+	void * __capability kernelinfo;
+};
+
+typedef struct aiocb_c {
+	int32_t	aio_fildes;		/* File descriptor */
+	off_t	aio_offset __packed;	/* File offset for I/O */
+	void * __capability aio_buf;		/* I/O buffer in process space */
+	size_t	 aio_nbytes;		/* Number of bytes for I/O */
+	int32_t	__spare__[2];
+	void * __capability  __spare2__;
+	int32_t	aio_lio_opcode;		/* LIO opcode */
+	int32_t	aio_reqprio;		/* Request priority -- ignored */
+	struct	__aiocb_private_c _aiocb_private;
+	struct	sigevent_c aio_sigevent;	/* Signal to deliver */
+} aiocb_c_t;
+
+static int
+aiocb_c_copyin(struct aiocb *ujob, struct aiocb *kjob)
+{
+	struct aiocb_c job_c;
+	int error;
+
+	error = copyincap(ujob, &job_c, sizeof(job_c));
+	if (error)
+		return (error);
+	CP(job_c, *kjob, aio_fildes);
+	CP(job_c, *kjob, aio_offset);
+	error = cheriabi_cap_to_ptr(__DEVOLATILE(caddr_t*, &kjob->aio_buf),
+	    job_c.aio_buf, job_c.aio_nbytes,
+	    /* XXX: Too strong.  Might not need store to write, load to read. */
+	    CHERI_PERM_GLOBAL|CHERI_PERM_STORE|CHERI_PERM_LOAD, 1);
+	if (error)
+		return (error);
+	CP(job_c, *kjob, aio_nbytes);
+	CP(job_c, *kjob, aio_lio_opcode);
+	CP(job_c, *kjob, aio_reqprio);
+	CP(job_c, *kjob, _aiocb_private.status);
+	CP(job_c, *kjob, _aiocb_private.error);
+	/* Nothing uses kernelinfo so we don't update it */
+	kjob->_aiocb_private.kernelinfo = 0;
+	return (convert_sigevent_c(&job_c.aio_sigevent, &kjob->aio_sigevent));
+}
+
+static long
+aiocb_c_fetch_status(struct aiocb *ujob)
+{
+	struct aiocb_c *ujob_c;
+
+	ujob_c = (struct aiocb_c *)ujob;
+	return (fuword(&ujob_c->_aiocb_private.status));
+}
+
+static long
+aiocb_c_fetch_error(struct aiocb *ujob)
+{
+	struct aiocb_c *ujob_c;
+
+	ujob_c = (struct aiocb_c *)ujob;
+	return (fuword(&ujob_c->_aiocb_private.error));
+}
+
+static int
+aiocb_c_store_status(struct aiocb *ujob, long status)
+{
+	struct aiocb_c *ujob_c;
+
+	ujob_c = (struct aiocb_c *)ujob;
+	return (suword(&ujob_c->_aiocb_private.status, status));
+}
+
+static int
+aiocb_c_store_error(struct aiocb *ujob, long error)
+{
+	struct aiocb_c *ujob_c;
+
+	ujob_c = (struct aiocb_c *)ujob;
+	return (suword(&ujob_c->_aiocb_private.error, error));
+}
+
+static int
+aiocb_c_store_kernelinfo(struct aiocb *ujob, long jobref)
```

Comment:
This is fine, I should probably just remove this upstream as it was never used anyway.